### DISCUSSION TITLE
GLTF (PBR) Shader

### DISCUSSION THREAD

**Discussion by ClonedPuppy** [StereoKit/StereoKit#502] [Q&A]
Hi, I'm working on optimising some things in my project, and up to now, I have specified my material passes (albedo, metal & rough) to my custom shader with `Mat[MatParamName.DiffuseTex] = Tex.FromFile(x.png)` etc.
I wanted to simplify this so that my custom shader just grabs the textures already included in the gltf file. This works fine in SK if I load in the gltf file without specifying a new material, however, when I assign my custom shader / material to it, I get no texture map data in the shader (shows just a grey model). I have used the code from SK's PBR shader itself that's in the /StereoKitC/shaders_builtin/ directory.
I also tried to just copy the SK PBR shader code and place it in a new file and assigned that, just to rule out any issues with my own shader, but no luck with that either. 

Up to now I figured SK auto assigned the default PBR shader to a gltf model, but perhaps that's a wrong assumption? I'm starting to think that SK might be parsing the gltf file first and then sending the textures to the shader under the hood ?

Thanks for any clarification!

**Comment by maluoi** [comment]
Yes, textures are applied to the `Material` only while loading the `Model`! They are not stored independently in the `Model`, so when applying a new `Material`, there's nothing automatic that will transfer textures over to the new `Material`, that has to be done manually in some form or another.

_Here's your options!_ (iiuc)

If you provide a `Shader` when calling `Model.FromFile`, StereoKit should fill that out as long as the shader uses the correct texture names! ("diffuse", "metal", "normal", "occlusion", "emission") The downside here is that SK uses the provided `Shader` for all the `Model`'s `Material`s regardless of what type they might originally result in. Not necessarily a blocker, but something to be aware of!

If you are assigning a completely new `Material` after the `Model` has been loaded, you need to get the existing `Material`'s texture values and assign them to the new one! `Material.GetTexture` was added in one of the v0.3.7 preview builds, so that's a somewhat recent thing.

I actually do have a `material_set_shader` on the native side of the API, not sure why it never made it into C#! But in theory that could allow you to `Material.Copy` and set a new `Shader`, preserving any overlapping `Shader` parameters. If that's of interest, let me know and I'll try and get it properly tested and added to the public API.

**Reply by ClonedPuppy** [reply]
Thank you for your reply, I have replied inline;

> If you provide a Shader when calling Model.FromFile, StereoKit should fill that out as long as the shader uses the correct texture names! ("diffuse", "metal", "normal", "occlusion", "emission") The downside here is that SK uses the provided Shader for all the Model's Materials regardless of what type they might originally result in. Not necessarily a blocker, but something to be aware of!

This works, as in the model do look correct, however once I set the material via: `model.Visuals[0].Material = MaterialWithNewShader`, it reverts back to no textures. I have to set this material since I'm sending data to the shader, I think there's no other way around this than to assign a material right?

> If you are assigning a completely new Material after the Model has been loaded, you need to get the existing Material's texture values and assign them to the new one! Material.GetTexture was added in one of the v0.3.7 preview builds, so that's a somewhat recent thing.

This works quite well, get's a bit verbose if many textures are used, but seems to do the trick. I use it like this, which I think is correct?
`material[MatParamName.DiffuseTex] = model.Visuals[0].Material.GetTexture("diffuse");`

> I actually do have a material_set_shader on the native side of the API, not sure why it never made it into C#! But in theory that could allow you to Material.Copy and set a new Shader, preserving any overlapping Shader parameters. If that's of interest, let me know and I'll try and get it properly tested and added to the public API.

I guess this might be the most elegant way? But I'm not sure how specific my usecase is, and if it's worth doing on your end, especially since GetTexture works.

**Comment by ClonedPuppy** [comment]
Been working a bit more on this, and I guess it could get a bit laboursome when dealing with other things that the material section hold in the gltf file. Just to get the most basic stuff over, takes quite a few Material.Get's. So yeah, perhaps it would be really useful to just be able to copy over all the values with a one-liner Material.Copy or something like that...

As an example, to just grab the non texture stuff from a model and move them to a custom shader, would require a "Get" for basecolor.rgba, metallicFactor and roughnessFactor... and then if the model has emissive etc it adds up quite a bit.
```
    "materials" : [
        {
            "doubleSided" : true,
            "name" : "Material",
            "pbrMetallicRoughness" : {
                "baseColorFactor" : [
                    0.1481827050447464,
                    0.1481827050447464,
                    0.1481827050447464,
                    1
                ],
                "metallicFactor" : 0.7545454502105713,
                "roughnessFactor" : 0
            }
        }
    ], 
```

**Reply by maluoi** [reply]
Okay, yeah, that sounds like an argument for the `SetShader` function! :) I added this to the list over here: https://github.com/StereoKit/StereoKit/issues/504

**Reply by maluoi** [reply]
I just went to implement `Material.SetShader()`, and realized there's already a `Material.Shader` that does exactly this! So you should be able to just use that :)

**Reply by ClonedPuppy** [reply]
Thanks!
